### PR TITLE
Write List, Patch and Delete LimitRange test +3 Endpoints

### LINKED ISSUE #112429: Write List, Patch and Delete LimitRange test +3 Endpoints
# Progress <code>[4/6]</code>

-   [X] APISnoop org-flow: [CoreV1LimitRangeTest.org](https://github.com/apisnoop/ticket-writing/blob/master/CoreV1LimitRangeTest.org)
-   [X] Test approval issue: [#112429](https://github.com/kubernetes/kubernetes/issues/112429) 
-   [X] Test PR: [#112430](https://github.com/kubernetes/kubernetes/pull/112430)  
-   [x] Two weeks soak start date: [testgrid-link](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=5&graph-metrics=test-duration-minutes&include-filter-by-regex=should.list,.patch.and.delete.a.LimitRange.by.collection) 4 Oct 2022
-   [ ] Two weeks soak end date: 18 Oct 2022
-   [ ] Test promotion PR: [#](https://pr.k8s.io/)

# Identifying an untested feature Using APISnoop

According to this APIsnoop query, there are still three remaining LimitRange endpoints which are untested.

```sql-mode
    SELECT
      endpoint,
      path,
      kind
      FROM testing.untested_stable_endpoint
      where eligible is true
      and endpoint ilike '%LimitRange%'
      order by kind, endpoint desc
      limit 10;
```

```example
                    endpoint                  |                       path                        |    kind
  --------------------------------------------+---------------------------------------------------+------------
   patchCoreV1NamespacedLimitRange            | /api/v1/namespaces/{namespace}/limitranges/{name} | LimitRange
   listCoreV1LimitRangeForAllNamespaces       | /api/v1/limitranges                               | LimitRange
   deleteCoreV1CollectionNamespacedLimitRange | /api/v1/namespaces/{namespace}/limitranges        | LimitRange
  (3 rows)

```

# API Reference and feature documentation

-   [Kubernetes API Reference Docs](https://kubernetes.io/docs/reference/kubernetes-api/)
-   [Kubernetes API / Policy Resource / LimitRange](https://kubernetes.io/docs/reference/kubernetes-api/policy-resources/limit-range-v1/)
-   [client-go - LimitRange](https://github.com/kubernetes/client-go/blob/master/kubernetes/typed/core/v1/limitrange.go)

# Test outline

```
Feature: Test list(All Namespaces), patch and deleteCollection LimitRange api endpoints
```

-   listCoreV1LimitRangeForAllNamespaces

```
Scenario: confirm that the list action for all namespaces will apply to a LimitRange
  Given the e2e test has created a LimitRange with a label
  When the test lists all LimitRanges with a label selector
  Then the requested action is accepted without any error
  And one LimitRange is found
```

-   patchCoreV1NamespacedLimitRange

```
Scenario: confirm that the patch action will apply the changes to a LimitRange
  Given the e2e test has a LimitRange after the "list" scenario
  And a valid payload has been created
  When the test patches the LimitRange
  Then the requested action is accepted without any error
  And the newly applied changes are found
```

-   deleteCoreV1CollectionNamespacedLimitRange

```
Scenario: confirm that the deleteCollection action will remove a LimitRange
  Given the e2e test has a LimitRange after the "patch" scenario
  When the test applies the deleteCollection action with a labelSelector
  Then the requested action is accepted without any error
  And the LimitRange with the label is not found
```

# E2E Test

Using a number of existing e2e test practices a new [ginkgo test](https://github.com/ii/kubernetes/blob/create-limitrange-test/test/e2e/scheduling/limit_range.go#L229-L311) has been created for three LimitRange endpoints. The e2e logs for this test are listed below.

```
[It] should ensure that a limitRange can be listed, patched and deleted by collection
  /home/heyste/go/src/k8s.io/kubernetes/test/e2e/scheduling/limit_range.go:229
STEP: Creating LimitRange "e2e-limitrange-s69c5" 09/14/22 10:47:30.685
STEP: Listing all LimitRanges with label "e2e-limitrange-s69c5=created" 09/14/22 10:47:30.711
Sep 14 10:47:30.716: INFO: Found limitRange "e2e-limitrange-s69c5" in namespace "limitrange-8238"
STEP: Patching LimitRange "e2e-limitrange-s69c5" 09/14/22 10:47:30.716
Sep 14 10:47:30.733: INFO: LimitRange "e2e-limitrange-s69c5" has been patched
STEP: Delete LimitRange "e2e-limitrange-s69c5" by Collection with labelSelector: "e2e-limitrange-s69c5=patched" 09/14/22 10:47:30.733
STEP: Confirm that the limitRange "e2e-limitrange-s69c5" has been deleted 09/14/22 10:47:30.782
Sep 14 10:47:30.783: INFO: Requesting list of LimitRange to confirm quantity
Sep 14 10:47:30.784: INFO: Found 0 LimitRange with label "e2e-limitrange-s69c5=patched"
Sep 14 10:47:30.784: INFO: LimitRange "e2e-limitrange-s69c5" has been deleted.
```

# Verifying increase in coverage with APISnoop

This query shows which LimitRange endpoints are hit within a short period of running this e2e test

```sql-mode
select distinct  substring(endpoint from '\w+') AS endpoint,
                 right(useragent,80) AS useragent
from testing.audit_event
where endpoint ilike '%LimitRange%'
and release_date::BIGINT > round(((EXTRACT(EPOCH FROM NOW()))::numeric)*1000,0) - 60000
and useragent like 'e2e%should%'
order by endpoint
limit 10;
```

```example
                  endpoint                  |                                    useragent
--------------------------------------------+----------------------------------------------------------------------------------
 createCoreV1NamespacedLimitRange           | should ensure that a limitRange can be listed, patched and deleted by collection
 deleteCoreV1CollectionNamespacedLimitRange | should ensure that a limitRange can be listed, patched and deleted by collection
 listCoreV1LimitRangeForAllNamespaces       | should ensure that a limitRange can be listed, patched and deleted by collection
 listCoreV1NamespacedLimitRange             | should ensure that a limitRange can be listed, patched and deleted by collection
 patchCoreV1NamespacedLimitRange            | should ensure that a limitRange can be listed, patched and deleted by collection
(5 rows)

```

# Final notes

If a test with these calls gets merged, **test coverage will go up by 3 point**

This test is also created with the goal of conformance promotion.

---

/sig testing
/sig architecture
/area conformance

### PR DESCRIPTION
**Pull request by riaankleinhans**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it:**
This PR adds a test to test the following untested endpoints:
- patchCoreV1NamespacedLimitRange
- listCoreV1LimitRangeForAllNamespaces
- deleteCoreV1CollectionNamespacedLimitRange

**Which issue(s) this PR fixes:**
Fixes #112429

**Testgrid Link:** 
[testgrid-link](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=5&graph-metrics=test-duration-minutes&include-filter-by-regex=should.list,.patch.and.delete.a.LimitRange.by.collection)

**Special notes for your reviewer:**
Adds +3 endpoint test coverage (good for conformance)

**Does this PR introduce a user-facing change?:**
```
NONE

```

**Release note:**
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:**
```
NONE

```

/sig testing
/sig architecture
/area conformance

### CODE REVIEW

File: test/e2e/scheduling/limit_range.go [hunk 1]
```diff
@@ -222,6 +226,89 @@ var _ = SIGDescribe("LimitRange", func() {
 		framework.ExpectNoError(err)
 	})
 
+	ginkgo.It("should ensure that a limitRange can be listed, patched and deleted by collection", func() {
```
Comment (kerthcet) [inline test/e2e/scheduling/limit_range.go:229]: Does this has something to do with scheduling? It seems like we're testing limitRange itself.

Comment (Riaankl) [reply test/e2e/scheduling/limit_range.go:229]: As part of addressing the [conformance technical debit](https://apisnoop.cncf.io/conformance-progress/endpoints/1.26.0/?filter=untested), the focus of this PR is to create an e2e test that will confirm the behaviour of the following limitRange endpoints.

- [listCoreV1LimitRangeForAllNamespaces](https://apisnoop.cncf.io/1.26.0/stable/core/listCoreV1LimitRangeForAllNamespaces?conformance-only=true)
- [patchCoreV1NamespacedLimitRange](https://apisnoop.cncf.io/1.26.0/stable/core/patchCoreV1NamespacedLimitRange?conformance-only=true)
- [deleteCoreV1CollectionNamespacedLimitRange](https://apisnoop.cncf.io/1.26.0/stable/core/deleteCoreV1CollectionNamespacedLimitRange?conformance-only=true)

The details of the reseach is captured in the issue [#112429](https://github.com/kubernetes/kubernetes/issues/112429).

Once the test proven stable (2 weeks flake free) we will promote it to conformance.

Comment (kerthcet) [reply test/e2e/scheduling/limit_range.go:229]: Of course, these tests are helpful. My concern here is do these tests have more or less relevances with pod scheduling, I didn't find any. If so, maybe put them under the package only related to limit range will be better, for maintainabilities. I don't know if we have one.

Comment (Riaankl) [reply test/e2e/scheduling/limit_range.go:229]: We placed the test in the limit_range.go file as it already contain the test for create -, delete -, list -, read -, replaceCoreV1NamespacedLimitRange endpoints, which is nested under test/e2e/scheduling/.
Moving the limit_range.go file to another location would likely have to be a SIG Scheduling decision and not related to the test.

Comment (alculquicondor) [inline test/e2e/scheduling/limit_range.go:229]: `It` refers to the feature (LimitRange), not the test. So it should be something like:

```golang
ginkgo.It("should list, patch and delete a LimitRange by collection")
```

File: test/e2e/scheduling/limit_range.go [hunk 2]
```diff
@@ -222,6 +226,89 @@ var _ = SIGDescribe("LimitRange", func() {
...
+		defaultLimit := getResourceList("500m", "500Mi", "500Gi")
+		defaultRequest := getResourceList("100m", "200Mi", "200Gi")
+		maxLimitRequestRatio := v1.ResourceList{}
+		value := strconv.Itoa(time.Now().Nanosecond()) + string(uuid.NewUUID())
+
+		limitRange := &v1.LimitRange{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: lrName,
+				Labels: map[string]string{
+					"time": value,
```
Comment (alculquicondor) [inline test/e2e/scheduling/limit_range.go:248]: what is this label for?

File: test/e2e/scheduling/limit_range.go [hunk 3]
```diff
@@ -222,6 +226,89 @@ var _ = SIGDescribe("LimitRange", func() {
...
+			},
+		}
+
+		ginkgo.By(fmt.Sprintf("Creating LimitRange %q", lrName))
+		limitRange, err := lrClient.Create(context.TODO(), limitRange, metav1.CreateOptions{})
+		framework.ExpectNoError(err, "Failed to create limitRange %q", lrName)
+
+		// Listing across all namespaces to verify api endpoint: listCoreV1LimitRangeForAllNamespaces
+		ginkgo.By(fmt.Sprintf("Listing all LimitRanges with label %q", createdLabelSelector))
+		limitRangeList, err := f.ClientSet.CoreV1().LimitRanges("").List(context.TODO(), metav1.ListOptions{LabelSelector: createdLabelSelector})
```
Comment (alculquicondor) [inline test/e2e/scheduling/limit_range.go:272]: If listing all namespaces, it's probably worth having a LR in at least 2 namespaces.

File: test/e2e/scheduling/limit_range.go (hunk 2 above)
Comment (heyste) [reply test/e2e/scheduling/limit_range.go:248]: Nothing now, was part of the initial code I was using to understand limitRange. Removing it.

**Comment by heyste** [comment]
@alculquicondor Is everything addressed in the latest update to the PR? TIA

File: test/e2e/scheduling/limit_range.go [hunk 4]
```diff
@@ -222,6 +226,103 @@ var _ = SIGDescribe("LimitRange", func() {
...
+					},
+				},
+			},
+		}
+		// Create a copy to be used in a second namespace
+		limitRange2 := &v1.LimitRange{}
+		*limitRange2 = *limitRange
+
+		ginkgo.By(fmt.Sprintf("Creating LimitRange %q in namespace %q", lrName, f.Namespace.Name))
+		limitRange, err := lrClient.Create(context.TODO(), limitRange, metav1.CreateOptions{})
```
Comment (alculquicondor) [inline test/e2e/scheduling/limit_range.go:269]: create a context at the beginning that you can use in all the calls.

File: test/e2e/scheduling/limit_range.go [hunk 5]
```diff
@@ -310,3 +411,22 @@ func newTestPod(name string, requests v1.ResourceList, limits v1.ResourceList) *
 		},
 	}
 }
+
+func checkLimitRangeListQuantity(f *framework.Framework, label string, quantity int) func() (bool, error) {
+	return func() (bool, error) {
+		var err error
```
Comment (alculquicondor) [inline test/e2e/scheduling/limit_range.go:417]: unnecessary line

Comment (heyste) [reply test/e2e/scheduling/limit_range.go]: Fixed

File: test/e2e/scheduling/limit_range.go (hunk 4 above)
Comment (heyste) [reply test/e2e/scheduling/limit_range.go]: Fixed

File: test/e2e/scheduling/limit_range.go [hunk 6]
```diff
@@ -222,6 +226,106 @@ var _ = SIGDescribe("LimitRange", func() {
...
+						MaxLimitRequestRatio: maxLimitRequestRatio,
+					},
+				},
+			},
+		}
+		// Create a copy to be used in a second namespace
+		limitRange2 := &v1.LimitRange{}
+		*limitRange2 = *limitRange
+
+		ctx, cancelCtx := context.WithTimeout(context.TODO(), wait.ForeverTestTimeout)
```
Comment (alculquicondor) [inline test/e2e/scheduling/limit_range.go:268]:
```suggestion
		ctx, cancelCtx := context.WithTimeout(context.Background(), wait.ForeverTestTimeout)
```

Comment (heyste) [reply test/e2e/scheduling/limit_range.go]: Thanks and fixed.

**Review by alculquicondor — COMMENTED** [review]
/approve

**Pull request merged**